File: src/abstract_factory.py
"""
Factory Method Pattern
======================

"""

from abc import ABC, abstractmethod


# Product
class Pizza(ABC):
    def __init__(self, name):
        self.name = name
        self.dough = None
        self.sauce = None
        self.veggies = []
        self.cheese = None
        self.pepperoni = None
        self.clam = None

    @abstractmethod
    def prepare(self): pass

    def bake(self):
        print("Bake for 25 minutes at 350")

    def cut(self):
        print("Cutting the pizza into diagonal slices")

    def box(self):
        print("Place pizza in official PizzaStore box")


class PizzaIngredientFactory(ABC):
    @abstractmethod
    def create_dough(self): pass

    @abstractmethod
    def create_sauce(self): pass

    @abstractmethod
    def create_cheese(self): pass

    @abstractmethod
    def create_veggies(self): pass

    @abstractmethod
    def create_pepperoni(self): pass

    @abstractmethod
    def create_clam(self): pass


class NYPizzaIngredientFactory(PizzaIngredientFactory):
    def create_dough(self): return "Thin Crust Dough"

    def create_sauce(self): return "Marinara Sauce"

    def create_cheese(self): return "Reggiano Cheese"

    def create_veggies(self): return ["Garlic", "Onion", "Mushroom", "RedPepper"]

    def create_pepperoni(self): return "Sliced Pepperoni"

    def create_clam(self): return "Fresh Clams"


class ChicagoIngredientFactory(PizzaIngredientFactory):
    def create_dough(self): return "Thick Crust Dough"

    def create_sauce(self): return "Plum Tomato Sauce"

    def create_cheese(self): return "Shredded Mozzarella Cheese"

    def create_veggies(self): return ["Spinach", "Black Olives", "Eggplant"]

    def create_pepperoni(self): return "Sliced Pepperoni"

    def create_clam(self): return "Frozen Clams"


class CheesePizza(Pizza):
    def __init__(self, ingredient_factory: PizzaIngredientFactory):
        super().__init__("Cheese Pizza")
        self.ingredient_factory = ingredient_factory

    def prepare(self):
        print("Preparing ", self.name)
        self.dough = self.ingredient_factory.create_dough()
        self.sauce = self.ingredient_factory.create_sauce()
        self.cheese = self.ingredient_factory.create_cheese()


class ClamPizza(Pizza):
    def __init__(self, ingredient_factory: PizzaIngredientFactory):
        super().__init__("Clam Pizza")
        self.ingredient_factory = ingredient_factory

    def prepare(self):
        print("Preparing ", self.name)
        self.dough = self.ingredient_factory.create_dough()
        self.sauce = self.ingredient_factory.create_sauce()
        self.cheese = self.ingredient_factory.create_cheese()
        self.clam = self.ingredient_factory.create_clam()


# Creator
class PizzaStore(ABC):
    @abstractmethod
    def create_pizza(self, pizza_type: str) -> Pizza:
        pass

    def order_pizza(self, pizza_type: str) -> Pizza:
        pizza = self.create_pizza(pizza_type)

        pizza.prepare()
        pizza.bake()
        pizza.cut()
        pizza.box()

        return pizza


# Concrete Creator
class NyPizzaStore(PizzaStore):
    def create_pizza(self, pizza_type: str) -> Pizza:
        pizza_type = pizza_type.lower()
        ingredient_factory = NYPizzaIngredientFactory()
        if pizza_type == "cheese":
            pizza = CheesePizza(ingredient_factory)
        elif pizza_type == "clam":
            pizza = ClamPizza(ingredient_factory)
        else:
            raise ValueError(f"Don't know how to create New York style {pizza_type} pizza")
        pizza.name = "Now York Style" + pizza.name
        return pizza


# Concrete Creator
class ChicagoPizzaStore(PizzaStore):
    def create_pizza(self, pizza_type: str) -> Pizza:
        pizza_type = pizza_type.lower()
        ingredient_factory = ChicagoIngredientFactory()
        if pizza_type == "cheese":
            pizza = CheesePizza(ingredient_factory)
        elif pizza_type == "clam":
            pizza = ClamPizza(ingredient_factory)
        else:
            raise ValueError(f"Don't know how to create Chicago style {pizza_type} pizza")
        pizza.name = "Chicago Style" + pizza.name
        return pizza


if __name__ == '__main__':
    ny_store = NyPizzaStore()
    chicago_store = ChicagoPizzaStore()

    pizza = ny_store.order_pizza("cheese")

    pizza = chicago_store.order_pizza("cheese")

File: src/abstract_factory.rs
//! Abstract Factory Pattern
//! ========================
//!

trait Pizza {
    fn set_name(&mut self, s: String);
    fn name(&self) -> &str;

    fn prepare(&mut self);

    fn bake(&mut self) {
        println!("Bake for 25 minutes at 350")
    }
    fn cut(&mut self) {
        println!("Cutting the pizza into diagonal slices")
    }
    fn boxify(&mut self) {
        println!("Place pizza in official PizzaStore box")
    }
}

trait PizzaIngredientFactory {
    fn create_dough(&self) -> &'static str;
    fn create_sauce(&self) -> &'static str;
    fn create_veggies(&self) -> Vec<&'static str>;
    fn create_cheese(&self) -> &'static str;
    fn create_pepperoni(&self) -> &'static str;
    fn create_clam(&self) -> &'static str;
}

trait PizzaStore {
    fn create_pizza(&self, pizza_type: &str) -> Box<dyn Pizza>;

    fn order_pizza(&self, pizza_type: &str) -> Box<dyn Pizza> {
        let mut pizza = self.create_pizza(pizza_type);

        pizza.prepare();
        pizza.bake();
        pizza.cut();
        pizza.boxify();

        pizza
    }
}

struct NyPizzaIngredientFactory;

impl PizzaIngredientFactory for NyPizzaIngredientFactory {
    fn create_dough(&self) -> &'static str {
        "Thin Crust Dough"
    }
    fn create_sauce(&self) -> &'static str {
        "Marinara Sauce"
    }
    fn create_veggies(&self) -> Vec<&'static str> {
        vec!["Garlic", "Onion", "Mushroom", "RedPepper"]
    }
    fn create_cheese(&self) -> &'static str {
        "Reggiano Cheese"
    }
    fn create_pepperoni(&self) -> &'static str {
        "Sliced Pepperoni"
    }
    fn create_clam(&self) -> &'static str {
        "Fresh Clams"
    }
}

struct ChicagoPizzaIngredientFactory;

impl PizzaIngredientFactory for ChicagoPizzaIngredientFactory {
    fn create_dough(&self) -> &'static str {
        "Thick Crust Dough"
    }
    fn create_sauce(&self) -> &'static str {
        "Plum Tomato Sauce"
    }
    fn create_veggies(&self) -> Vec<&'static str> {
        vec!["Spinach", "Black Olives", "Eggplant"]
    }
    fn create_cheese(&self) -> &'static str {
        "Shredded Mozzarella Cheese"
    }
    fn create_pepperoni(&self) -> &'static str {
        "Sliced Pepperoni"
    }
    fn create_clam(&self) -> &'static str {
        "Frozen Clams"
    }
}

struct CheesePizza {
    name: String,
    ingredient_factory: Box<dyn PizzaIngredientFactory>,
}

impl CheesePizza {
    pub fn new(ingredient_factory: impl 'static + PizzaIngredientFactory) -> Self {
        Self {
            name: "Cheese Pizza".to_string(),
            ingredient_factory: Box::new(ingredient_factory),
        }
    }
}

impl Pizza for CheesePizza {
    fn set_name(&mut self, name: String) {
        self.name = name;
    }

    fn name(&self) -> &str {
        &self.name
    }

    fn prepare(&mut self) {
        println!("Preparing {}", self.name);
        println!("  kneading {}", self.ingredient_factory.create_dough());
        println!("  smearing {}", self.ingredient_factory.create_sauce());
        println!("  spreading {}", self.ingredient_factory.create_cheese());
    }
}

struct ClamPizza {
    name: String,
    ingredient_factory: Box<dyn PizzaIngredientFactory>,
}

impl ClamPizza {
    pub fn new(ingredient_factory: impl 'static + PizzaIngredientFactory) -> Self {
        Self {
            name: "Clam Pizza".to_string(),
            ingredient_factory: Box::new(ingredient_factory),
        }
    }
}

impl Pizza for ClamPizza {
    fn set_name(&mut self, name: String) {
        self.name = name;
    }

    fn name(&self) -> &str {
        &self.name
    }

    fn prepare(&mut self) {
        println!("Preparing {}", self.name);
        println!("  kneading {}", self.ingredient_factory.create_dough());
        println!("  smearing {}", self.ingredient_factory.create_sauce());
        println!("  spreading {}", self.ingredient_factory.create_cheese());
        println!("  placing {}", self.ingredient_factory.create_clam());
    }
}

struct NyPizzaStore;

impl PizzaStore for NyPizzaStore {
    fn create_pizza(&self, pizza_type: &str) -> Box<dyn Pizza> {
        let ingredient_factory = NyPizzaIngredientFactory;
        let mut pizza: Box<dyn Pizza> = match pizza_type {
            "cheese" => Box::new(CheesePizza::new(ingredient_factory)),
            "clam" => Box::new(ClamPizza::new(ingredient_factory)),
            _ => panic!(
                "Don't know how to create New York style {} pizza",
                pizza_type
            ),
        };
        pizza.set_name(format!("New York Style {}", pizza.name()));
        pizza
    }
}

struct ChicagoPizzaStore;

impl PizzaStore for ChicagoPizzaStore {
    fn create_pizza(&self, pizza_type: &str) -> Box<dyn Pizza> {
        let ingredient_factory = ChicagoPizzaIngredientFactory;
        let mut pizza: Box<dyn Pizza> = match pizza_type {
            "cheese" => Box::new(CheesePizza::new(ingredient_factory)),
            "clam" => Box::new(ClamPizza::new(ingredient_factory)),
            _ => panic!(
                "Don't know how to create Chicago style {} pizza",
                pizza_type
            ),
        };
        pizza.set_name(format!("New Chicago Style {}", pizza.name()));
        pizza
    }
}

#[test]
fn test_run() {
    let ny_store = NyPizzaStore;
    let chicago_store = ChicagoPizzaStore;

    ny_store.order_pizza("cheese");
    chicago_store.order_pizza("clam");
}
